File: cPlusPlusCode/sobelFilter.cpp
#include<opencv2/imgproc.hpp>
#include<opencv2/highgui.hpp>
#include<iostream>

using namespace cv;
using namespace std;

int main(){

    //Read image
    Mat image = imread("../assets/putin.jpg", IMREAD_GRAYSCALE);

    //check if image is empty
    if(image.empty()){
        cout<<"can not find image"<<endl;
        return 0;
    }

    Mat gradientX, gradientY;

    //apply sobel filter for X gradient
    Sobel(image, gradientX, CV_32F, 1, 0);

    //apply sobel filter for Y gradient
    Sobel(image, gradientY, CV_32F, 0, 1);

    //normalize gradient X and gradient Y to display images
    normalize(gradientX, gradientX, 0, 1, NORM_MINMAX);
    normalize(gradientY, gradientY, 0, 1, NORM_MINMAX);

    //create windows to display image
    namedWindow("image", WINDOW_NORMAL);
    namedWindow("imagex", WINDOW_NORMAL);
    namedWindow("imagey", WINDOW_NORMAL);

    //display images
    imshow("image", image);
    imshow("imagex", gradientX);
    imshow("imagey", gradientY);

    //press esc to exit program
    waitKey(0);

    //close all the opened windows
    destroyAllWindows();
    return 0;
}

File: pythonCode/sobelFilter.py
import cv2
import sys
import numpy as np

#Read image
image = cv2.imread("../assets/putin.jpg", cv2.IMREAD_GRAYSCALE)

#check if image exits
if image is None:
    print("can not find image")
    sys.exit()

#apply sobel filter along X direction
gradientX = cv2.Sobel(image, cv2.CV_32F, 1, 0)

#apply sobel filter along Y direction
gradientY = cv2.Sobel(image, cv2.CV_32F, 0, 1)

#normalize images to display
cv2.normalize(gradientX, dst=gradientX, alpha=0, beta=1, norm_type=cv2.NORM_MINMAX, dtype=cv2.CV_32F)
cv2.normalize(gradientY, dst=gradientY, alpha=0, beta=1, norm_type=cv2.NORM_MINMAX, dtype=cv2.CV_32F)

#create windows to display images
cv2.namedWindow("image", cv2.WINDOW_NORMAL)
cv2.namedWindow("imagex", cv2.WINDOW_NORMAL)
cv2.namedWindow("imagey", cv2.WINDOW_NORMAL)

#display images
cv2.imshow("image", image)
cv2.imshow("imagex", gradientX)
cv2.imshow("imagey", gradientY)

#press esc to exit the program
cv2.waitKey(0)

#close all the opened windows
cv2.destroyAllWindows()